In [1]:
# For the Dataset Iris

import numpy as np
from sklearn.datasets import load_iris
iris = load_iris()

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(iris['data'],iris['target'], random_state=186)

Imported IRIS dataset here and splitted it using the train_test_split function and selected the random state as my birthdate.

In [2]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(112, 4) (38, 4) (112,) (38,)


In [3]:
# Calculating Euclidean Distance

def euc_distance(x,y) :
    euc_calc = np.empty(len(x))
    for i in range(len(x)) :
        euc_calc[i] =(x[i] - y[i]) ** 2
    answer = np.sum(euc_calc) ** 0.5
    return(answer)

In [4]:
# now i am using function min_finder that will give us the minimum value from array/list

import math
def min_finder(x):
    
    min = math.inf
    for i in range (len(X)):
        if X[i] < min:
            min = X[i]
    return min


In [5]:
# now using function argmin_finder to get the index value of the minimum value which we got previously

def argmin_finder(X):
    min = math.inf
    argmin = 0
    for i in range (len(X)):
        if X[i] < min:
            min = X[i]
            argmin = i
    return argmin



In [6]:
#calculating the nearest distance by creating 

distance = np.empty(len(X_test)*len(X_train)).reshape(len(X_test),len(X_train))
for i in range(len(X_test)):
    for j in range(len(X_train)):
        distance[i,j] = euc_distance(X_test[i],X_train[j])
        
        

y_pred = np.empty(len(X_test))
for i in range(len(X_test)):
    y_pred[i] = y_train[argmin_finder(distance[i])]

In [7]:
error_count = sum(y_pred!=y_test)
error_rate = (error_count/len(X_test))
error_rate

0.02631578947368421

After loading and splitting the dataset successfully first calcluated the euclidean distance after that using the min_finder function calculated the minimum value from the whole array.

After using the min_finder function, used the argmin_finder for getting the index value of that particular minimum value. lastly created the one empty set to get the distance between test and train sample and stored that into the y_pred to get the final error rate for the 1 nearest neighbour.

I have got the error rate as 0.0263 and accuracy as 0.973 that is 97%.

In [8]:
# 3NN algorithm for the dataset IRIS

def argmin_finder_3(x):
    
    min = math.inf
    x_c=[i for i in x]
    argmin = [0,1,2]
    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[0]=argmin1
    del(x_c[argmin1])
    min = math.inf

    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[1]=argmin1
    del(x_c[argmin1])
    min = math.inf

    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[2]=argmin1
    return argmin

Here for getting error rate for 3 nearest neighbours firstly i am finding the indexes for the 3 nearest points lets say for 0,1 & 2 using the min function we set it as to the infinity.
using the for loop we are trying to get the minimum value for that first data point (means in the first for loop its for data point 'zero'). after getting the least value we deleting it from the loop and only selecting the value which has the most number of occurances.
Again repeating the same process for the remaining data points as well.

In [9]:
y_pred_3 = np.empty(len(X_test))
for i in range(len(X_test)):
    pred_index=argmin_finder_3(distance[i])
    s=set(y_train)
    dict1={}
    for v in s:
        dict1[v]=0
    for j in pred_index:
        dict1[y_train[j]]+=1
    l=[0]
    for k1,v1 in dict1.items():
        l.append(v1)
    l.sort()
    
    if(l[-1]==1):
        y_pred_3[i] = y_train[pred_index[0]]
        break
    for k2,v2 in dict1.items():
        if(v2==l[-1]):
            y_pred_3[i] = k2
            
            break
            
       

In this y_pred_3 creating the empty set for storing the index values for each of the test data(labels).
creating the dictonary so that we can give each label a key because i will be using the same logic for the ionospehere dstaset as well.
At the end we are incrementing the value by one unit with its key values so that we can get the label which has the most number of occurances and then at the end we just selecting a label which ahs the most number of occurances. with the help of append we are storing the highest getting occurances.

Got the same accuracy and error rate as gor for 1 nearest neighbours.

In [10]:
error_count = sum(y_pred_3!=y_test)
error_rate = (error_count/len(X_test))
error_rate

0.02631578947368421

In [16]:
# for the dataset IONOSPEHRE


import numpy as np
ionosphere_data = np.genfromtxt('ionosphere.txt', delimiter=',', usecols=np.arange(34))
ionosphere_target = np.genfromtxt('ionosphere.txt', delimiter=',', usecols=34)

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(ionosphere_data, ionosphere_target, random_state=186)

Downloaded and uploaded Ionosphere dataset here and splitted it using the train_test_split function and selected the random state as my birthdate.

In [18]:
print(X_train.shape, X_test.shape, y_test.shape, y_train.shape)

(263, 34) (88, 34) (88,) (263,)


In [19]:
import numpy as np
def euc_distance(x,y) :
    euc_calc = np.empty(len(x))
    for i in range(len(x)) :
        euc_calc[i] =(x[i] - y[i]) ** 2
    answer = np.sum(euc_calc) ** 0.5
    return(answer)

import math
def min_finder(x):
    
    min = math.inf
    for i in range (len(X)):
        if X[i] < min:
            min = X[i]
    return min

In [20]:
def argmin_finder(X):
    min = math.inf
    argmin = 0
    for i in range (len(X)):
        if X[i] < min:
            min = X[i]
            argmin = i
    return argmin


In [21]:
distance = np.empty(len(X_test)*len(X_train)).reshape(len(X_test),len(X_train))
for i in range(len(X_test)):
    for j in range(len(X_train)):
        distance[i,j] = euc_distance(X_test[i],X_train[j])
        
        

y_pred = np.empty(len(X_test))
for i in range(len(X_test)):
    y_pred[i] = y_train[argmin_finder(distance[i])]

In [22]:
error_count = sum(y_pred!=y_test)
error_rate = (error_count/len(X_test))
error_rate

0.11363636363636363

After loading and splitting the dataset successfully first calcluated the euclidean distance after that using the min_finder function calculated the minimum value from the whole array.

After using the min_finder function, used the argmin_finder for getting the index value of that particular minimum value. lastly created the one empty set to get the distance between test and train sample and stored that into the y_pred to get the final error rate for the 1 nearest neighbour.

I have got the error rate as 0.113 and accuracy as 0.89 that is 89%.

In [24]:
# 3NN for the Ionospehre

def argmin_finder_3(x):
    
    min = math.inf
    x_c=[i for i in x]
    argmin = [0,1,2]
    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[0]=argmin1
    del(x_c[argmin1])
    min = math.inf

    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[1]=argmin1
    del(x_c[argmin1])
    min = math.inf

    for i in range(len(x_c)):
        if x_c[i] < min:
            min = x_c[i]
            argmin1 = i
    argmin[2]=argmin1
    return argmin

In [25]:
y_pred_3 = np.empty(len(X_test))
for i in range(len(X_test)):
    pred_index=argmin_finder_3(distances[i])
    s=set(y_train)
    dict1={}
    for v in s:
        dict1[v]=0
    for j in pred_index:
        dict1[y_train[j]]+=1
    l=[0]
    for k1,v1 in dict1.items():
        l.append(v1)
    l.sort()
    
    if(l[-1]==1):
        y_pred_3[i] = y_train[pred_index[0]]
        break
    for k2,v2 in dict1.items():
        if(v2==l[-1]):
            y_pred_3[i] = k2
            
            break

In [26]:
error_count = sum(y_pred_3!=y_test)
error_rate = (error_count/len(X_test))
error_rate

0.19318181818181818

In this y_pred_3 creating the empty set for storing the index values for each of the test data(labels).
creating the dictonary so that we can give each label a key because i will be using the same logic for the ionospehere dstaset as well.
At the end we are incrementing the value by one unit with its key values so that we can get the label which has the most number of occurances and then at the end we just selecting a label which ahs the most number of occurances. with the help of append we are storing the highest getting occurances.

Got the different error rate accuracy for the 3nn for the ionospehere dataset.
Error rate=0.19 and accuracy=81%